hangman_game.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Hangman.Game do
+  defmodule State do
+    defstruct(game_state: ":initializing", 
+              turns_left: 7,               
+              shown_letters: [],        #If user won or lost, letters list would be shown_letters list
+              letters: [],
+              used: [],
+              word: "",
+              last_guess: "" 
+              )
+  end
+
+  def tally(state = %State{letters: l, shown_letters: sl}) when (l == sl) do
+      %{game_state: ":won", 
+        turns_left: state.turns_left, 
+        letters: state.shown_letters,
+        used: [], 
+        last_guess: state.last_guess
+        }
+  end
+
+  def tally(state = %State{turns_left: tl}) when (tl == 0) do
+      %{game_state: ":lost", 
+        turns_left: state.turns_left, 
+        letters: state.shown_letters,
+        used: [], 
+        last_guess: ""
+        }
+  end
+
+  def tally(state = %State{}) do               
+      %{game_state: state.game_state, 
+             turns_left: state.turns_left, 
+             letters: state.letters,
+             used: state.used, 
+             last_guess: state.last_guess
+             }
+  end
+
+  def new_game() do
+      word = Dictionary.random_word()
+      shown_letters = String.codepoints(word)
+      letters = Enum.map(shown_letters, fn(_) -> "_" end)
+      %State{shown_letters: shown_letters, letters: letters, word: word}
+  end
+
+  def make_move(current_state, try_letter) do
+      current_state = %State{current_state | last_guess: try_letter}
+      new_state = handle_guess(jackpot?(current_state, try_letter),
+                                try_letter_used?(current_state, try_letter),
+                                current_state
+                                )
+      {new_state, tally(new_state)}
+  end    
+
+  def handle_guess([], false, current_state) do      #When try fail and letter is not in used.
+      %State{current_state | game_state: ":bad_guess", 
+                        turns_left: current_state.turns_left - 1,
+                        used: Enum.sort(current_state.used ++ [current_state.last_guess])
```

Comment:
## -1 idiomatic

Rather than writing 

~~~ elixir
current_state.used ++ [current_state.last_guess]
~~~

use

~~~ elixir
[ current_state.last_guess | current_state.used ]
~~~